src/api/c/vector_field.cpp

```diff
@@ -50,25 +51,23 @@ fg_chart setup_vector_field(fg_window window, const vector<af_array>& points,
     vector<Array<T>> pnts;
     vector<Array<T>> dirs;
 
+    dim_t ndots = 0;
     for (unsigned i = 0; i < points.size(); ++i) {
         pnts.push_back(getArray<T>(points[i]));
         dirs.push_back(getArray<T>(directions[i]));
+        ndots += getInfo(points[i]).dims().dims[0];
     }
 
     // Join for set up vector
-    dim4 odims(3, points.size());
-    Array<T> out_pnts = createEmptyArray<T>(odims);
-    Array<T> out_dirs = createEmptyArray<T>(odims);
-    detail::join(out_pnts, 1, pnts);
-    detail::join(out_dirs, 1, dirs);
-    Array<T> pIn = out_pnts;
-    Array<T> dIn = out_dirs;
-
-    // do transpose if required
-    if (transpose_) {
-        pIn = transpose<T>(pIn, false);
-        dIn = transpose<T>(dIn, false);
-    }
+    dim4 odims(transpose_ ? ndots : 3, transpose_ ? 3 : ndots);
+    // initialize the full out_ptns & out_dirs arrays, because join of 2D points
+    // will only partially fill those arrays
+    Array<T> out_pnts = createValueArray<T>(odims, 0.);
+    Array<T> out_dirs = createValueArray<T>(odims, 0.);
```

Comment:
Since dimensions are similar, I don't see the need for `createValueArray`, it adds an unnecessary mem set call.

Reply:
I had the same thought when looking into the problem, although will debugging I discovered the unexpected usage.

We have 4 entry scenario's (and I use the quantities of field example)
1. transpose=true
   a) pnts.dims()=[1156,2,1,1], dirs.dims()=[1156,2,1,1]  --> odims.T()=[3,1156,1,1]  --> field example
   b) pnts.dims()=[1156,3,1,1], dirs.dims()=[1156,3,1,1]  --> odims.T()=[3,1156,1,1]
2. transpose=false
   a) pnts.dims()=[2,1156,1,1], dirs.dims()=[2,1156,1,1]  --> odims=[3,1156,1,1]
   b) pnts.dims()=[3,1156,1,1], dirs.dims()=[3,1156,1,1]  --> odims=[3,1156,1,1]
PS: For multiple array points, the output array will by [3,sum of all points of all input arrays, 1,1]

For 2D inputs + transponse, the join will copy X & Y and leaving the Z uninitialized.
   input array storage (X11,X12,...,X1n, Y11,Y12,...,Y1n) & (X21,X22,...,X2, Y21,Y22,...,Y2n) & ...
       --> odims      (X11,X12,...X1n,X21,X22,..X2n,... , Y11,Y12,..Y1n,Y21,Y22,..Y2n,..., ?,?,?,?...)  !!Z-axis uninitialized!!
       --> odims.T() (X11,Y11,?, X12,Y12,?, ...)
For 3D inputs the full result array will be written.


As alternative to pre-initializing the full output array, we can also add the constant array of 0.  I did not do so, because I wanted to remain as close to possible to the original code.


PS: The original calculation of odims took the number of arrays iso the total number of points.  This could generate a memory overflow on the GPU buffer.

Reply:
Your previous statement, did make me doubt so I returned to a very old version of arrayfire 3.7.2, where I saw that the conversion to 3D wasn't done neither although It displayed a better result.

Finally, only the dimensions were wrong, resulting in a possible buffer overrun on the GPU.

So I willy push a new version, with lesser changes although with better results.